geps/gep-1651.md

```diff
@@ -0,0 +1,274 @@
+Gateways using the same GatewayClass on the cluster, but ensure that the
+"internal" service is only reachable within the cluster. This would greatly
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of
+layer 7 load balancing for our internal services.
+
+## API
+
+We propose adding a new `reachability` field under the `spec.infrastructure` stanza of a Gateway.
+
+### Predefined Reachability Values
+
+Implementations MAY implement the following values for 'reachability' and MUST abide by
+their defined semantics.
+
+Value | Semantic
+-|-
+`public`|The address is reachable over the public internet
+`cluster`|The address is only reachability within the bounds defined by the [cluster's network model](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+
+### Vendor prefixed values
+
+Implementations can define custom 'reachability' values by specifying a vendor prefix followed
+by a slash `/` and a custom name ie. `com.example.com/my-reachability`.
+
+### Default reachability
+The default value of `reachability` is implementation specific. It is RECOMMENDED that
+the default `reachability` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+### Mutability
+Implementations MAY prevent end-users from updating the `reachability` value of a Gateway. If
+updates are allowed the semantics and behaviour will depend on the underlying implementation.
+
+### Go
+
+```go
+
+type GatewayReachability string
+
+const (
+  GatewayReachabilityPublic  GatewayReachability = "public"
+  GatewayReachabilityCluster GatewayReachability = "cluster"
+)
+
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
+  Infrastructure GatewayInfrastructure `json:"infrastructure"`
+  // ...
+}
+type GatewayInfrastructure struct {
+  // AddressScope allows the Gateway to specify the accessibility of it's addresseses
+  Reachability GatewayReachability `json:"reachability"`
+}
+```
+
+### YAML
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  infrastructure:
+    reachability: public
+  listeners:
+  - protocol: HTTP
+    port: 80
+```
+
+### Semantics
+
+#### Interaction with GatewayClass
+
+An infrastructure provider may provide a pre-defined set of GatewayClasses that limit the
+reachability choices of a Gateway. If the desired Gateway reachability is incompatible with the
+GatewayClass it MUST set the condition `Accepted` to `False` with `Reason` set to `Invalid`.
+
+#### Unsupported reachability & address values
+
+If a Gateway is unable provide an address for the desired reachability it MUST set the condition `Accepted` to `False` with `Reason` set to `Invalid`
+
+#### Status.Addresses
```

Comment:
Can we add a `reachability` field to the status.address and have it reported there as well?

Reply:
Done - please take another look.